clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -280,7 +295,27 @@ void Sema::checkSYCLDeviceVarDecl(VarDecl *Var) {
   SourceRange Loc = Var->getLocation();
   llvm::DenseSet<QualType> Visited;
 
-  checkSYCLVarType(*this, Ty, Loc, Visited);
+  checkSYCLVarType(*this, Ty, Loc, Visited, false);
+}
+
+static void checkAccessorVarType(Sema &S, QualType Ty, SourceRange Loc) {
+  assert(Util::isSyclAccessorType(Ty) &&
+         "Should only be called on SYCL accessor types.");
+
+  llvm::DenseSet<QualType> Visited;
+  const RecordDecl *RecTy = Ty->getAsRecordDecl();
+  for (const auto &Field : RecTy->fields()) {
+    if (Field->getName().compare("MData") == 0) {
+      QualType FTy = Field->getType();
+      // Sub-reference array or pointer, then proceed with that type.
+      while (FTy->isAnyPointerType() || FTy->isArrayType())
+        FTy = QualType{FTy->getPointeeOrArrayElementType(), 0};
+      FTy = FTy.getUnqualifiedType().getCanonicalType();
+
+      checkSYCLVarType(S, FTy, Loc, Visited, true);
```

Comment:
I feel that checkSYCL**Var**Type is not really good name since you don't check the type of **variable** here.

Reply:
That function already existed before this PR.  It's primary caller is `Sema::checkSYCLDeviceVarDecl` .  I'd be happy to rename it. Do you have a suggestion?

Reply:
I'm not really good in naming, but how about just `checkSYCLType`.

Reply:
done